daemon/attach.go

```diff
@@ -114,131 +113,96 @@ func (daemon *Daemon) ContainerAttach(job *engine.Job) engine.Status {
+				log.Debugf("attach: stdin: end")
 			}()
-		}
-	}
-	if stdout != nil {
-		nJobs++
-		// Get a reader end of a pipe that is attached as stdout to the container.
-		if p, err := streamConfig.StdoutPipe(); err != nil {
-			errors <- err
-		} else {
-			cStdout = p
-			go func() {
-				log.Debugf("attach: stdout: begin")
-				defer log.Debugf("attach: stdout: end")
-				// If we are in StdinOnce mode, then close stdin
-				if stdinOnce && stdin != nil {
-					defer stdin.Close()
-				}
-				_, err := io.Copy(stdout, cStdout)
-				if err == io.ErrClosedPipe {
-					err = nil
-				}
-				if err != nil {
-					log.Errorf("attach: stdout: %s", err)
-				}
-				errors <- err
-			}()
-		}
-	} else {
-		// Point stdout of container to a no-op writer.
-		go func() {
-			if cStdout, err := streamConfig.StdoutPipe(); err != nil {
-				log.Errorf("attach: stdout pipe: %s", err)
+			var err error
+			if tty {
+				_, err = utils.CopyEscapable(cStdin, stdin)
 			} else {
-				io.Copy(&ioutils.NopWriter{}, cStdout)
+				_, err = io.Copy(cStdin, stdin)
+
 			}
-		}()
-	}
-	if stderr != nil {
-		nJobs++
-		if p, err := streamConfig.StderrPipe(); err != nil {
-			errors <- err
-		} else {
-			cStderr = p
-			go func() {
-				log.Debugf("attach: stderr: begin")
-				defer log.Debugf("attach: stderr: end")
-				// If we are in StdinOnce mode, then close stdin
-				// Why are we closing stdin here and above while handling stdout?
-				if stdinOnce && stdin != nil {
-					defer stdin.Close()
-				}
-				_, err := io.Copy(stderr, cStderr)
-				if err == io.ErrClosedPipe {
-					err = nil
-				}
-				if err != nil {
-					log.Errorf("attach: stderr: %s", err)
-				}
-				errors <- err
-			}()
-		}
-	} else {
-		// Point stderr at a no-op writer.
-		go func() {
-			if cStderr, err := streamConfig.StderrPipe(); err != nil {
-				log.Errorf("attach: stdout pipe: %s", err)
-			} else {
-				io.Copy(&ioutils.NopWriter{}, cStderr)
+			if err == io.ErrClosedPipe {
+				err = nil
+			}
+			if err != nil {
+				log.Errorf("attach: stdin: %s", err)
 			}
+			errors <- err
 		}()
+	} else {
+		errors <- nil
 	}
 
-	return promise.Go(func() error {
+	attachStream := func(name string, stream io.Writer, streamPipe io.ReadCloser) {
+		if stream == nil {
+			errors <- nil
+			return
+		}
 		defer func() {
-			if cStdout != nil {
-				cStdout.Close()
-			}
-			if cStderr != nil {
-				cStderr.Close()
+			// Make sure stdin gets closed
+			if stdinOnce && cStdin != nil {
+				stdin.Close()
+				cStdin.Close()
 			}
+			streamPipe.Close()
 		}()
 
+		log.Debugf("attach: %s: begin", name)
+		defer log.Debugf("attach: %s: end", name)
+		_, err := io.Copy(stream, streamPipe)
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: %s: %v", name, err)
+		}
+		errors <- err
+	}
+
+	go attachStream("stdout", stdout, cStdout)
+	go attachStream("stderr", stderr, cStderr)
+
+	return promise.Go(func() error {
 		for i := 0; i < nJobs; i++ {
+			err := <-errors
```

Comment:
Now debug message about waiting is after really waiting. I don't know why you changed this at all :)